hilti/toolchain/include/ast/type.h

```diff
@@ -2,414 +2,582 @@
+
+} // namespace type
+
+/** * Base class for classes implementing unqualified types. */
+class UnqualifiedType : public Node {
 public:
-    Flags() = default;
-    Flags(Flag f) : _flags(static_cast<uint64_t>(f)) {}
-    Flags(const Flags&) = default;
-    Flags(Flags&&) noexcept = default;
-    ~Flags() = default;
-
-    /** Returns true if a given flag has been set. */
-    bool has(Flag f) const { return _flags & static_cast<uint64_t>(f); }
-
-    /** Sets (or clear) a given flag. */
-    void set(type::Flag flag, bool set = true) {
-        if ( set )
-            _flags |= static_cast<uint64_t>(flag);
-        else
-            _flags &= ~static_cast<uint64_t>(flag);
-    }
+    ~UnqualifiedType() override;
 
-    Flags operator+(Flag f) {
-        auto x = Flags(*this);
-        x.set(f);
-        return x;
-    }
+    /**
+     * Returns the index the AST context associates with the type. This may
+     * become set during AST resolving. If not set yet, returns `None`.
+     */
+    auto typeIndex() const { return _type_index; }
 
-    Flags operator+(const Flags& other) const {
-        auto x = Flags();
-        x._flags = _flags | other._flags;
-        return x;
-    }
+    /**
+     * Returns the index the AST context associates with the declaration
+     * declaring this type. This may become set during AST resolving. If not
+     * set yet, returns `None`.
+     */
+    auto declarationIndex() const { return _declaration_index; }
 
-    Flags& operator+=(Flag f) {
-        set(f);
-        return *this;
-    }
-    Flags& operator+=(const Flags& other) {
-        _flags |= other._flags;
-        return *this;
+    /**
+     * Sets the type's declaration index that the context maintains. Note that
+     * this does not update any state inside the context, so ensure setting it
+     * to something valid.
+     *
+     * @param index index to set
+     */
+    void setDeclarationIndex(ast::DeclarationIndex index) {
+        assert(index);
+        _declaration_index = index;
     }
 
-    Flags operator-(const Flags& other) const {
-        auto x = Flags();
-        x._flags = _flags & ~other._flags;
-        return x;
-    }
+    /**
+     * Returns the declaration declaring this type, or null if none. This is a
+     * shortcut for retrieving the `decclarationIndex()` and then looking it up
+     * in the AST context. That means, the index must have been set for this to
+     * return anything.
+     */
+    std::shared_ptr<declaration::Type> typeDeclaration() const;
 
-    Flags& operator-=(Flag f) {
-        set(f, false);
-        return *this;
-    }
-    Flags& operator-=(const Flags& other) {
-        _flags &= ~other._flags;
-        return *this;
-    }
+    /**
+     * Returns the C++ ID associated with this type, if any. This is a shortcut
+     * to retrieving the associated declaration's C++ ID, which is set through
+     * a `&cxxname` attribute.
+     */
+    ID cxxID() const;
```

Comment:
All these getters for an `ID` need to return a copy since we want to allow semantically unset `ID`s. Do you think it would make sense to return e.g., `optional_ref<ID>` instead?